Refactor CLI

watersofoblivion opened the pull request:
This revamps the CLI implementation to be cleaner and more idiomatic.

# Configuration

The configuration now uses package variables.  These are a little easier to read IMO and are better documented.  It also eliminates passing derefed pointers as arguments to functions.

Flag parsing has been moved to a separate top-level function which puts flag parsing and (almost) all error checking in one location.

The command name and the additional args are now parsed immediately after the flags and put in package variables as well so that the rest of the code can use them easily.

# Commands

The command names have been extracted out into constants and the implementation of each command has been moved into a standalone function.  The makes the command loop more obvious and easy to follow.  It also makes the implementation of each command more clear.

Additionally, some of the local variable names have been cleaned up and unnecessary intermediate variables removed. 

# Output

The usage of `fmt` has been eliminated and replaced with `log`.  Since `log` prints a timestamp with each message, the need for the hand-rolled timer has been eliminated.  This does lose the "Overall" time being printed out at the end, but it allows users to see how long each migration took or how long it ran before erroring out.  The overall time can be eyeballed from looking at the first and last timestamp so it's not truly lost, just not explicitly stated.

The several instances of 

```go
fmt.Print(msg)
os.Exit(1)
```

have been replaced with a more idiomatic `log.Fatal(msg)`

It is important to note that the change from `fmt` to `log` does change the output from going to STDOUT to STDERR.  If this is likely to break clients, it is possible to add a call to `log.SetOutput(os.Stdout)` in an `init` function to mitigate the effect.

Also, several instances of unnecessary intermediate variables (mostly around color output) were removed and several redundant casts in `writePipe` were eliminated.

mattes commented:
Cool! Thanks for the PR! I will look into it today or tomorrow.

mattes commented:
My thoughts below. Happy to discuss. 

* I don't want to get rid of the overall timer. Can we please bring this back?
* Output should go to stdout, unless it's an error. 
* I'd rather merge `init()`, `Configure()`, and `ConfigureArgs()` into the `main()`. 
* I prefer the big `case` stmt better actually, but I'm open for discussion. The `const` declarations at the top just seem to add boilerplate code as this package won't be imported anywhere anyway.
* I like the idea of adding timestamps for each migration a lot.

watersofoblivion commented:
I've added the overall timer back in.  I've also DRY-ed up the `case` statement, addressed below.

For output to STDOUT vs. STDERR, while I have a preference in the end I'm fine with either one, but both is probably not what you'd want.  First, splitting between the two as you suggest would be slightly awkward in the codebase.  I'd have to set the output to STDOUT and then manually set the output to STDERR before each call to `log.Fatal()`.  Second, splitting between the two somewhat breaks being able to pipe the output to something like `grep` or redirecting to a file which is common in automation (and this is likely to be automated as part of a deploy.)  My preference is to follow the example of the `go` tool and print to STDERR.  It produces no output on success, on failure writes to STDERR, and when the `-v` flag is set writes all debugging output to STDERR.  This sends human-readable messages to STDERR while reserving STDOUT for printing things that are likely to be processed as part of a pipe chain.

I think it's cleaner to have the main read like "Configure the app, then process the commands" as is makes it easier for someone to jump in and follow the overall logic before having to parse through the implementation details.  To that end:

For merging the `init()`, `Configure()`, and `ConfigureArgs()` into `main()`, how about merging the first three into one `Configure()` function and calling that from `main()`?

Having the implementation of the various commands split into stand-alone functions also helps with this.  It makes it clearer that the `case` statement is just dispatching and keeps the `main()` function under a page so it's viewable as a whole.

watersofoblivion commented:
Went ahead and merged the configuration into a single function cleaned up the command dispatching enough to eliminate a few of the trivial command functions.

mattes commented:
> I'd have to set the output to STDOUT and then manually set the output to STDERR before each call to log.Fatal().

We could use two loggers.

```
stdLog := log.New(os. Stdout, "", log.Lshortfile)
stdErr := log.New(os. Stderr, "", log.Lshortfile)
```

And `fmt` has `Fprintln` which allows to specify a io.Writer, too.

> Second, splitting between the two somewhat breaks being able to pipe the output to something like grep or redirecting to a file

You can just redirect or merge the file descriptors. 

At the moment, the stdout/stderr is inconsistent. I believe we should reserve stderr for actual errors though. Many CI systems highlight stderr red.

Will go over the latest changes later this afternoon.

mattes commented:
Hi @watersofoblivion, sorry for the delay of my reply. I will close this PR as it is not adding new features and I don't want to break existing code. We should consider using package `log` for version 2 with a clear strategy on what is being logged to stdout/stderr. Thanks for your efforts.
